extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+              partitionOffsets.startingOffset,
+              partition
+          );
+        }
+      } else { // subsequent run, use metadata storage to check consistency
+        partitionOffsets.startingOffset = partitionOffsets.endingOffset;
+        if (storedOffsets.get(partition) == null || storedOffsets.get(partition) != partitionOffsets.endingOffset) {
+          log.error(
+              "endingOffset from metadata storage [%d] does not match expected offset [%d] for partition [%d]",
+              (storedOffsets.get(partition) == null ? -1 : storedOffsets.get(partition)),
+              partitionOffsets.endingOffset,
+              partition
+          );
+        }
+      }
+
+      partitionOffsets.endingOffset = partitionOffsets.startingOffset + messagesPerPartition;
+      partitionOffsets.active = true;
+    }
+  }
+
+  /**
+   * Compares the list of partitions/offsets for taskId from TaskStorage and the data stored in taskGroup.partitions
+   * (which is the current set of partitions/offsets being processed) and returns false if these do not match.
+   */
+  private boolean isTaskCurrent(int taskGroupId, String taskId)
+  {
+    TaskGroup taskGroup = taskGroups.get(taskGroupId);
+    if (!taskStorage.getTask(taskId).isPresent() || taskGroup == null) {
+      return false;
+    }
+
+    KafkaIndexTask task = (KafkaIndexTask) taskStorage.getTask(taskId).get();
+    Map<Integer, Long> taskStartPartitions = task.getIOConfig().getStartPartitions().getPartitionOffsetMap();
+    Map<Integer, Long> taskEndPartitions = task.getIOConfig().getEndPartitions().getPartitionOffsetMap();
+
+    for (Integer partition : taskStartPartitions.keySet()) {
+      if (taskGroup.partitions.get(partition) == null
+          || taskGroup.partitions.get(partition).startingOffset != taskStartPartitions.get(partition)
+          || taskGroup.partitions.get(partition).endingOffset != taskEndPartitions.get(partition)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private void killRunningTasksInGroup(int taskGroupId)
+  {
+    if (taskMaster.getTaskQueue().isPresent()) {
+      Map<String, Optional<TaskStatus>> taskStatuses = taskGroups.get(taskGroupId).taskStatuses;
+      for (String taskId : taskStatuses.keySet()) {
+        if (taskStatuses.get(taskId).isPresent() && !taskStatuses.get(taskId).get().isComplete()) {
+          taskMaster.getTaskQueue().get().shutdown(taskId);
+        }
+      }
+    } else {
+      log.error("Failed to get task queue because I'm not the leader!");
+    }
+  }
+
+  private void killUnownedKafkaTasks()
+  {
+    if (taskMaster.getTaskQueue().isPresent()) {
+      List<Task> tasks = taskStorage.getActiveTasks();
+      for (Task task : tasks) {
+        if (task instanceof KafkaIndexTask
+            && dataSource.equals(task.getDataSource())
+            && !supervisorRunId.equals(task.getContextValue(SUPERVISOR_CONTEXT_KEY))) {
```

Comment:
Hmm yeah, the 'killing any task that wasn't started by us' behavior came out of discussions of the possibility of two overlords both thinking they are the leader and spawning a bunch of Kafka indexing tasks that do the same thing. I think the concern was that there'd be a scenario where they didn't generate the same sequenceNames (somehow were using different starting offset ranges or a different ingestion spec) and then we'd get data duplication, so if I remember correctly we thought it might make sense for new supervisors to just kill the other overlord's tasks until the split-leader state settled. 

I like the idea of taking a more intelligent look at the running tasks and seeing if they're doing what we want and only taking action if they don't match our expectation.